### PR TITLE
fix #22622: Partial fix, export barlines subtypes

### LINKED ISSUE #22622: We must stop using the "Offset" property to determine the default position of items
### Issue type

Engraving bug (incorrect score rendering)

### Summary 

What follows is just an explanation and a collection of the most obvious issues which arise from this. I am sure parts of this have been reported before. All of these problems come back to one single original design error: 

**The "Offset" property must be reserved for the _user's_ offset. It must _not_ be used for the default positions of the program.**

This is something which I've encountered over and over again. It has now surfaced yet again as a blocker for the new ["Center between staves" feature](https://github.com/musescore/MuseScore/issues/22175), unless I find I way to circumvent the issue. I'm logging this now as an issue of its own, so it's there for when we can get to it. It will be a big project, not only because we will need to rewrite all the default positions of all the items without using offsets, but also because we have years and years of scores that were generated this way, so we'll have to do migrations etc. 

## Explanation and examples

In Musescore, the default position of many items is determined by pre-applying an offset to them. See for instance:
image

Essentially, we have a concept of "default offset". This is a _disaster_, and it keeps causing an insane amount of engraving issues and general UX jank (which corresponds to code jank and hacks and workarounds underneath).

1. The concept of "default offset" is in itself contradictory. You offset something _with respect to a default_, or a reference point. A "default offset" is... not an offset. It is just the default position of the item on the score.
2. From a UX point of view, it is confusing and inconsistent. If I click on other items, like notes, I find their default position to have _zero_ offset. So I (correctly) interpret the "Offset" property to represent the _user's_ chosen position with respect to the default one. But when I select the hairpin, like above, I'm contradicted, and I have to guess that in that case the offset is meant _with respect to the bottom of the staff_. So which is it?
image

3. The "default offset" is meant to determine the default position. But the default position is _also_ determined by the autoplace system. Then consider this: what does a 2sp offset mean in this case? (Spoiler: nothing, it's absolutely meaningless). 
image

4. The "Offset" property is still supposed to hold the user's customized movement, so we have _one_ single variable (the offset) representing _both_ a default position _and_ a user position.  Let's say I want that hairpin to be slightly further away from the note. So I enter 3sp in the vertical offset field and... nothing happens. Why? Because the autoplace has already moved that hairpin to a point that's further than 3sp from the staff.
image

5. When mouse-dragging an item, the dragging operation has to move the item with respect to the position that the user sees on screen. But that position was calculated by autoplace. But autoplace already took an offset into account (the default one). But now I'm dragging so I'm also adding new offset. This can (and does, often) cause things to "explode" when dragging items, cause it's easy to run into infinite loops. To avoid this, we do a hack of "rebasing" the offset. "Rebasing" means that we add to the offset value the movement that had already been done by the autoplace. So, as soon as I start draggins this hairpin, the offset value jumps from 2sp to 5sp, even though I haven't even moved it yet.
image
Then, say, I want to put the hairpin back to the default position. I would imagine that that's what I'd get if I'd reset the offset.
image
But no: because we're now in a "offset rebased" state, the offset will indeed go back to 2sp, but meant as the original "2sp below the staff", which means we get this.
image

6. Notice from the previous example that, when the hairpin ends up overlapping the note, the "Minimum distance" property has suddenly been updated to an inexplicable negative value. That's not a bug, it's done intentionally. Why? Because, since the offset is also used to determine default positions, the autoplace system must take offsets into account, ending up with a situation where the offset of an item affects the autoplace of itself. Say, in the next example, I need to save space, so I move that dynamic inside the staff. I do that by dragging (i.e. applying an offset to) it. If we left "Minimum distance" as normal, the autoplace would take my offset into account and therefore "offset my offset" in order to prevent the collision with the staff, effectively disallowing me to move the dynamic inside the staff. To circumvent this problem, we hack the minimum distance and set it to a large negative value, so that the autoplace will not interpret that as a collision and let us do what we want.
image

### Steps to reproduce

Just try using the offset property on basically any item for more than 30 seconds.

### Screenshots/Screen recordings

_No response_

### MuseScore Version

Current master

### Regression

No.

### Operating system

Any

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by akhisud**
This fixes it partially. It exports the barline styles correclty- i.e, double, dotted, dashed etc but doesn't handle staffs spanned.

### CODE REVIEW

File: mscore/exportxml.cpp [hunk 1]
```diff
@@ -5102,6 +5158,7 @@ void ExportMusicXml::write(QIODevice* dev)
                                     spannerStart(this, strack, etrack, st, sstaff, seg);
                                     }
 
+
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: whitespace damage

File: mscore/exportxml.cpp [hunk 2]
```diff
@@ -5123,7 +5180,8 @@ void ExportMusicXml::write(QIODevice* dev)
                         repeatAtMeasureStop(xml, m, strack, etrack, strack);
                   // note: don't use "m->repeatFlags() & Repeat::END" here, because more
                   // barline types need to be handled besides repeat end ("light-heavy")
-                  barlineRight(m);
+                  //barlineRight(m);
+
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp [hunk 3]
```diff
@@ -5392,3 +5450,4 @@ void ExportMusicXml::harmony(Harmony const* const h, FretDiagram const* const fd
 
 }
 
+
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp [hunk 4]
```diff
@@ -4839,6 +4839,62 @@ void ExportMusicXml::writeElement(Element* el, const Measure* m, int sstaff, boo
                   // TODO: print barline only if middle
                   // if (el->subtype() != BarLineType::START_REPEAT)
                   //       bar((BarLine*) el);
+
+                  if(el->parent()==m->last()){
+                      const Measure* mmR1 = m->mmRest1(); // the multi measure rest this measure is covered by
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: please indent by 6 spaces

File: mscore/exportxml.cpp [hunk 5]
```diff
@@ -4839,6 +4839,62 @@ void ExportMusicXml::writeElement(Element* el, const Measure* m, int sstaff, boo
...
+                      const Measure* mmRLst = mmR1->isMMRest() ? mmR1->mmRestLast() : 0; // last measure of replaced sequence of empty measures
+                      // note: use barlinetype as found in multi measure rest for last measure of replaced sequence
+                      BarLineType bst = m == mmRLst ? mmR1->endBarLineType() : m->endBarLineType();
+                      BarLine* bl =  static_cast<BarLine*>(el);
+                      if(bl->customSubtype())
+                          bst= bl->barLineType();
```
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: ditto

File: mscore/exportxml.cpp (hunk 4 above)
Comment (Jojo-Schmitz) [inline mscore/exportxml.cpp]: please indent by 6 spaces

**Comment by Jojo-Schmitz** [comment]
Travis failed and seemling for good reasons, in mtest for  tst_midimapping. Either the tests need to get adjusted or the code, I'm not sure which

**Comment by Jojo-Schmitz** [comment]
Why? You could have fixed the issues instead?

**Comment by akhisud** [comment]
I'll fix them and reopen the PR

**Comment by Jojo-Schmitz** [comment]
OK, but you could also without closing the PR.

**Comment by akhisud** [comment]
Is there a lint code that I could just run on my program so that it fixes spaces etc according to the rules for MuseScore coding?

**Comment by akhisud** [comment]
I have a feeling there's a problem with the tests that are making the checks fail above. I believe travis has a wrong (old?) version of the reference files for which it is failing.  It says some lines should be there and some not, which is not the case in the reference test files in the current repo. Can this be confirmed?
Namely, these files: https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest1_ref.xml
https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest2_ref.xml
https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest3_ref.xml

**Comment by akhisud** [comment]
I have a feeling there's a problem with the tests that are making the checks fail above. I believe travis has a wrong (old?) version of the reference files for which it is failing. It says some lines should be there and some not, which is not the case in the reference test files in the current repo. Can this be confirmed?
Namely, these files: https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest1_ref.xml
https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest2_ref.xml
https://github.com/musescore/MuseScore/blob/master/test/musicxml/testMultiMeasureRest3_ref.xml

**Comment by Jojo-Schmitz** [comment]
See https://musescore.org/en/developers-handbook/musescore-coding-rules there's a settings file for QtCreator that takes care of the coding style
@lvinken would be the guy to do a tech review

**Comment by lvinken** [comment]
Sorry, but the auto test on Travis is correct and your changes really do not pass: testMultiMeasureRest3 fails and testSystemBrackets1 segfaults.

Please verify the auto test runs OK before submitting a pull request.
Tip: on Unix run "(cd mtest/musicxml/io/; ./iotest)" at the top-level of the source tree.

**Comment by akhisud** [comment]
The latest build on my travis (https://travis-ci.org/akhisud/MuseScore/jobs/119016323) gives this:

46/46 Test #41: tst_mxml_io ......................***Exception: Other 11.80 sec

as the only problematic part. What does this mean?

**Comment by Jojo-Schmitz** [comment]
I guess that is the segvailt @lvinken talked about

**Comment by akhisud** [comment]
I don't think it is. I was able to resolve the segfault that it threw out

**Comment by akhisud** [comment]
The following tests FAILED:
	 41 - tst_mxml_io (OTHER_FAULT).

**Comment by Jojo-Schmitz** [comment]
It could also be a temp glitch on Travis and pass on next attempt

**Comment by AntonioBL** [comment]
The first failing test is TestMxmlIO::systemBrackets1(), related to file testSystemBrackets1.xml in the mtest\musicxml\io folder.
Attached the backtrace for this test. Apparently, a null element is being accessed as the result of retrieving track 4 element from lastSegment (which only has a bar line at track 0).
[backtrace.txt](https://github.com/musescore/MuseScore/files/193229/backtrace.txt)

**Comment by akhisud** [comment]
@AntonioBL, thanks, that was the problem after all!  Could you tell me how you identified this particular file as giving the problem? I could only see tst_mxml_io.

**Comment by Jojo-Schmitz** [comment]
to be continued in #2503

**Comment by AntonioBL** [comment]
@akhisud : well, I simply run the tests locally and inside gdb, just to be sure to catch the error, i.e (assuming a Linux OS):
cd MuseScore
make debug       (to compile MuseScore as a debug build)
cd build.debug/mtest/musicxml/io    (to compile MusicXML I/O tests)
make
gdb ./tst_mxml_io
(and then "run")

Ciao,
ABL

**Comment by akhisud** [comment]
@AntonioBL , Thanks for the tip and the debug :)

**Pull request closed without merging**